### PR TITLE
Add EOF1 container support in libevmasm

### PR DESCRIPTION
**Pull request by axic**
Part of #13717.

### CODE REVIEW

**Comment by axic** [comment]
Unfortunately now this old version of code is not enough, because need to adjust jump offsets. They need to be within the code section, i.e. in this simplified PR we could just start all jumps with a `+10` offset (due to the fixed header size).

**Comment by axic** [comment]
Need to expose `eofVersion` in `AbstractAssembly` and passed down to `assemble()`.

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -539,14 +539,36 @@ LinkerObject const& Assembly::assemble() const
...
+	// Insert EOF1 header.
+	bytesRef eofCodeLength(&ret.bytecode.back(), 0);
+	bytesRef eofDataLength(&ret.bytecode.back(), 0);
+	if (eof)
+	{
+		// TODO: empty data is disallowed
+		ret.bytecode.push_back(0xef);
+		ret.bytecode.push_back(0x00);
+		ret.bytecode.push_back(0x01); // version 1
+		ret.bytecode.push_back(0x01); // kind=code
```
Comment (axic) [inline libevmasm/Assembly.cpp:559]: This header is fine for single-code section case. For functions it needs to be generalised and needs a type section too.

File: libevmasm/Assembly.h [hunk 2]
```diff
@@ -114,7 +114,8 @@ class Assembly
 	langutil::SourceLocation const& currentSourceLocation() const { return m_currentSourceLocation; }
 
 	/// Assembles the assembly into bytecode. The assembly should not be modified after this call, since the assembled version is cached.
-	LinkerObject const& assemble() const;
+	/// @param eof If true, assemble for EOF, otherwise for legacy EVM output.
+	LinkerObject const& assemble(bool eof = false) const;
```
Comment (axic) [inline libevmasm/Assembly.h]: Need to change ~~`CompilerStack`~~ libyul to pass this flag down.

File: libevmasm/Assembly.cpp [hunk 3]
```diff
@@ -774,10 +774,15 @@ LinkerObject const& Assembly::assemble() const
 		toBigEndian(ret.bytecode.size(), r);
 	}
 
-	auto const dataLength = ret.bytecode.size() - dataStart;
+	auto dataLength = ret.bytecode.size() - dataStart;
 	if (m_eofVersion.has_value())
 	{
-		assertThrow(/*dataLength >= 0 && */ dataLength <= 0xffff, AssemblyException, "Invalid data section size.");
+		if (dataLength == 0)
```
Comment (axic) [inline libevmasm/Assembly.cpp:781]: Should add a comment here, this should be temporary 😅

File: test/libsolidity/SolidityExecutionFramework.h [hunk 4]
```diff
@@ -62,6 +62,37 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
 	) override
 	{
 		bytes bytecode = multiSourceCompileContract(_sourceCode, _sourceName, _contractName, _libraryAddresses);
+
+		if (bytecode.size() > 0)
+		{
+			if (bytecode[0] == uint8_t(0xef)) // is EOF
```
Comment (axic) [inline test/libsolidity/SolidityExecutionFramework.h]: I know this is testing only, but
```suggestion
			if (bytecode[0] == uint8_t(0xef) && bytecode[1] == uint8_t(00)) // is EOF
```

Plus the `bytecode.size() > 3` makes more sense.

Maybe also add a TODO note to validate version at some point.

File: test/libsolidity/SolidityExecutionFramework.h [hunk 5]
```diff
@@ -62,6 +62,39 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
 	) override
 	{
 		bytes bytecode = multiSourceCompileContract(_sourceCode, _sourceName, _contractName, _libraryAddresses);
+
+		// Note: Temporary solution that adds arguments size in the EOF data_section_size header to validate evmone implementation.
+		// TODO: Validate header and version.
+		if (bytecode.size() > 3)
+		{
+			if (bytecode[0] == uint8_t(0xef) && bytecode[1] == uint8_t(00)) // is EOF
```
Comment (nikola-matic) [inline test/libsolidity/SolidityExecutionFramework.h:70]:
```suggestion
			if (bytecode[0] == uint8_t(0xef) && bytecode[1] == uint8_t(0x00)) // is EOF
```
Is this a typo? I mean, they're equivalent anyway, but whatever :D

**Comment by nikola-matic** [comment]
@axic @r0qs @ekpyron is this supposed to end up being merge-able, or is it just a PoC before the spec is finalized?

File: test/libsolidity/SolidityExecutionFramework.h [hunk 6]
```diff
@@ -62,6 +62,39 @@ class SolidityExecutionFramework: public solidity::test::ExecutionFramework
...
+			{
+				const auto firstSectionOffset = 3; // magic + version
+				bool stop = false;
+				for (size_t i = firstSectionOffset; i < bytecode.size(); ++i)
```
Comment (nikola-matic) [inline test/libsolidity/SolidityExecutionFramework.h:74]:
```suggestion
				for (size_t i = firstSectionOffset; i < bytecode.size() && !stop; ++i)
```
Should allow for getting rid of
```
if (stop)
    break;
```

**Comment by ekpyron** [comment]
> @axic @r0qs @ekpyron is this supposed to end up being merge-able, or is it just a PoC before the spec is finalized?

In general, the sooner we get this stuff actually merge-able the better - but first of all it serves as PoC (and as basis for the other EOF-PoCs). But this PR here doesn't even have the most recent version of this - that'd be here:

https://github.com/ethereum/solidity/compare/eof1-container...eof1-container-test

Which also adds a CI run that actually tests this stuff. But those tests still fail, since the evmone branch implementing EOF still doesn't implement the most recent spec of things :-).

So yeah, in practice, I don't see an imminent merge happening here :-).

**Comment by axic** [comment]
@ekpyron I'd vote getting all those commits into this PR and preparing this for merge. If there's any header change it can be applied. It will be worse maintaining all these separate branches/PRs imho.

**Comment by ekpyron** [comment]
@axic do you have a branch of evmone without the to-be-dropped requirement for terminating instructions at the end? Without that we can't merge this anyways.

**Comment by axic** [comment]
@ekpyron I think that can be sorted out, but yeah you're right this depends on testing suite, which is the achilles heel.

**Comment by ekpyron** [comment]
If we get such an evmone branch, though, we can indeed set somebody on making this mergable.

**Comment by NunoFilipeSantos** [comment]
Closing in favor of https://github.com/ethereum/solidity/pull/13825.

**Pull request closed without merging**